compiler_gym/service/proto/py_converters.py

```diff
@@ -2,151 +2,798 @@
+    res = res.reshape(tensor.shape)
+    return res
 
-        def compose_choice_dict(action: DictType[str, Any]):
-            return Action(
-                choice=[
-                    choice.make_choice(action[choice.space.name]) for choice in choices
-                ]
+
+def convert_numpy_to_boolean_tensor_message(tensor: np.ndarray):
+    return BooleanTensor(values=tensor.flatten().tolist(), shape=tensor.shape)
+
+
+def convert_byte_tensor_message_to_numpy(tensor: ByteTensor):
+    res = np.frombuffer(tensor.values, dtype=np.byte)
+    res = res.reshape(tensor.shape)
+    return res
+
+
+def convert_numpy_to_byte_tensor_message(tensor: np.ndarray):
+    return ByteTensor(values=tensor.tobytes(), shape=tensor.shape)
+
+
+def convert_numpy_to_int64_tensor_message(tensor: np.ndarray):
+    return Int64Tensor(values=tensor.flatten(), shape=tensor.shape)
+
+
+def convert_numpy_to_float_tensor_message(tensor: np.ndarray):
+    return FloatTensor(values=tensor.flatten(), shape=tensor.shape)
+
+
+def convert_numpy_to_double_tensor_message(tensor: np.ndarray):
+    return DoubleTensor(values=tensor.flatten(), shape=tensor.shape)
+
+
+def convert_numpy_to_string_tensor_message(tensor: np.ndarray):
+    return StringTensor(values=tensor.flatten(), shape=tensor.shape)
+
+
+convert_tensor_message_to_numpy = TypeBasedConverter(
+    conversion_map={
+        BooleanTensor: convert_standard_tensor_message_to_numpy,
+        ByteTensor: convert_byte_tensor_message_to_numpy,
+        Int64Tensor: convert_standard_tensor_message_to_numpy,
+        FloatTensor: convert_standard_tensor_message_to_numpy,
+        DoubleTensor: convert_standard_tensor_message_to_numpy,
+        StringTensor: convert_standard_tensor_message_to_numpy,
+    }
+)
+
+
+def convert_bytes_to_numpy(arr: bytes) -> np.ndarray:
+    return np.frombuffer(arr, dtype=np.int8)
+
+
+class NumpyToTensorMessageConverter:
+    dtype_conversion_map: DictType[Type, Callable[[Any], Message]]
+
+    def __init__(self):
+        self.dtype_conversion_map = {
+            np.bool_: convert_numpy_to_boolean_tensor_message,
+            np.int8: convert_numpy_to_byte_tensor_message,
+            np.int64: convert_numpy_to_int64_tensor_message,
+            np.float32: convert_numpy_to_float_tensor_message,
+            np.float64: convert_numpy_to_double_tensor_message,
+            np.dtype(object): convert_numpy_to_string_tensor_message,
+        }
+
+    def __call__(
+        self, tensor: np.ndarray
+    ) -> Union[
+        BooleanTensor, ByteTensor, Int64Tensor, FloatTensor, DoubleTensor, StringTensor
+    ]:
+        return self.dtype_conversion_map[tensor.dtype.type](tensor)
+
+
+convert_numpy_to_tensor_message = NumpyToTensorMessageConverter()
+
+
+def convert_trivial(val: Any):
+    return val
+
+
+# Convert a protobuf message to an object.
+# The conversion function is chosen based on the message descriptor.
+class FromMessageConverter:
```

Comment:
Nit, use docstrings:

```suggestion
class FromMessageConverter:
    """Convert a protobuf message to an object.

    The conversion function is chosen based on the message descriptor.
    """
```

Reply:
Done.